### PR TITLE
[mt] Implement vector intercept.

### PR DESCRIPTION
**Pull request by trivialfis**
- Extract the fit intercept procedures from the learner.
- Implement vector transform in objectives.
- Implement estimation for softmax objective.
- Remove scalar intercept. It will always return a vector.
- A lot of changes due to the new base score storage type.

todo:
- [x] python tests
- [x] c++ tests for utilities.
- [x] sklearn, load attributes, coef.
- [x] test weighted histogram.

Ref: https://github.com/dmlc/xgboost/issues/9043
Close https://github.com/dmlc/xgboost/issues/11374

Some PRs were extracted from this one:
- https://github.com/dmlc/xgboost/pull/11663
- https://github.com/dmlc/xgboost/pull/11658

### CODE REVIEW

File: src/objective/quantile_obj.cu [hunk 1]
```diff
@@ -165,20 +156,10 @@ class QuantileRegression : public ObjFunction {
...
-    std::array<double, 2> dat{meanq, sw};
-    auto rc = collective::GlobalSum(ctx_, info, linalg::MakeVec(dat.data(), dat.size()));
+    auto rc = collective::GlobalSum(ctx_, info, base_score->View(this->ctx_->Device()));
     collective::SafeColl(rc);
-
-    std::tie(meanq, sw) = std::tuple_cat(dat);
-    meanq /= (sw + kRtEps);
-    base_score->Reshape(1);
-    base_score->Data()->Fill(meanq);
+    double n_workers = info.IsColumnSplit() ? 1.0 : collective::GetWorldSize();
```
Comment (trivialfis) [inline src/objective/quantile_obj.cu:164]: We should still consider the weight for distributed mean

File: tests/cpp/test_learner.cc [hunk 2]
```diff
@@ -640,7 +660,12 @@ class TestColumnSplit : public ::testing::TestWithParam<std::string> {
 
  public:
   void Run(std::string objective) {
-    auto p_fmat = MakeFmatForObjTest(objective, 10, 10);
+    std::shared_ptr<DMatrix> p_fmat;
```
Comment (trivialfis) [inline tests/cpp/test_learner.cc]: Need a way to clean up the data gen

File: src/objective/multiclass_obj.cu [hunk 3]
```diff
@@ -167,29 +183,98 @@ class SoftmaxMultiClassObj : public ObjFunction {
...
+    auto intercept = base_score->View(ctx_->Device());
+    CHECK_EQ(intercept.Size(), n_classes);
+    CHECK_EQ(n, info.num_row_);
+
+    if (ctx_->IsCPU()) {
+      for (std::size_t i = 0; i < n; ++i) {
+        auto y = labels(i);
+        auto w = weights[i];
+        intercept(static_cast<std::size_t>(y)) += w;
+        sum_weight += w;
```
Comment (trivialfis) [inline src/objective/multiclass_obj.cu]: Skip sum weight

**Comment by trivialfis** [comment]
cc @rongou
